platforms/nuttx/src/px4/common/gpio/mcp23009/mcp23009.cpp

```diff
@@ -31,182 +31,133 @@
  *
  ****************************************************************************/
 
+#include <nuttx/ioexpander/gpio.h>
 #include <drivers/drv_sensor.h>
-#include <px4_platform/gpio/mcp23009.hpp>
-#include "mcp23009_registers.hpp"
-
-using namespace Microchip_MCP23009;
-
-const struct gpio_operations_s MCP23009::gpio_ops = {
-go_read : MCP23009::go_read,
-go_write : MCP23009::go_write,
-go_attach : nullptr,
-go_enable : nullptr,
-go_setpintype : MCP23009::go_setpintype,
+#include <lib/drivers/device/Device.hpp>
+#include <uORB/topics/gpio_config.h>
+#include <uORB/topics/gpio_in.h>
+#include <uORB/topics/gpio_out.h>
+#include <uORB/topics/gpio_request.h>
+#include <uORB/Publication.hpp>
+#include <uORB/SubscriptionCallback.hpp>
+
+static uint32_t DEVID{0};
+struct mcp23009_gpio_dev_s {
+	struct gpio_dev_s gpio;
+	uint8_t mask;
 };
 
-MCP23009::MCP23009(int bus, int address, int first_minor, int bus_frequency) :
-	I2C(DRV_GPIO_DEVTYPE_MCP23009, "MCP23009", bus, address, bus_frequency),
-	_first_minor(first_minor)
+/* Copy the read input data */
+class ReadCallback : public uORB::SubscriptionCallback
 {
-}
+public:
+	using SubscriptionCallback::SubscriptionCallback;
+
+	void call() override
+	{
+		px4::msg::GpioIn new_input;
+		if (update(&new_input) && new_input.device_id == DEVID) {
+			input = new_input.state;
+		}
 
-MCP23009::~MCP23009()
-{
-	/* set all as input & unregister */
-	for (int i = 0; i < num_gpios; ++i) {
-		go_setpintype(i, GPIO_INPUT_PIN);
-		gpio_pin_unregister(&_gpio[i].gpio, _first_minor + i);
 	}
-}
 
-int MCP23009::go_read(struct gpio_dev_s *dev, bool *value)
-{
-	mcp23009_gpio_dev_s *gpio = (struct mcp23009_gpio_dev_s *)dev;
-	return gpio->obj->go_read(gpio->id, value);
-}
+	uint8_t input;
+};
 
-int MCP23009::go_write(struct gpio_dev_s *dev, bool value)
+static uORB::Publication<px4::msg::GpioRequest> toGpioRequest{ORB_ID(gpio_request)};
+static ReadCallback fromGpioIn{ORB_ID(gpio_in)};
+static int mcp23009_read(struct gpio_dev_s *dev, bool *value)
 {
 	mcp23009_gpio_dev_s *gpio = (struct mcp23009_gpio_dev_s *)dev;
-	return gpio->obj->go_write(gpio->id, value);
+	*value = fromGpioIn.input & gpio->mask;
+	return OK;
```

Comment:
So the read now relies on the driver periodically reading the GPIO and notifying this interface of the mask. It's… UNBLOCKABLE, muhoha!